### PR TITLE
feat(contracts): integrate token-based permissions

### PR DESCRIPTION
**Pull request by kangsorang**
# Description

Modified the membership of the security council, which was managed as a separate list, to be based on token.

### CODE REVIEW

File: packages/contracts/contracts/universal/SimpleMultiSigWallet.sol [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+    /**
+     * @inheritdoc ISimpleMultiSigWallet
+     */
+    function submitTransaction(
+        address _target,
+        uint256 _value,
+        bytes memory _data
+    ) public onlyGovernance(msg.sender) returns (uint256) {
+        return _submitTransaction(_target, _value, _data);
```
Comment (0xHansLee) [inline packages/contracts/contracts/universal/SimpleMultiSigWallet.sol:108]: Why don't we add auto-confirm when submitting transaction? With this, we don't need to confirm transaction in `SecurityCouncil.sol:requestDeletion()`
```suggestion
        uint256 transactionId = _submitTransaction(_target, _value, _data);
        confirmTransaction(transactionId);
        return transactionId;
```

File: packages/contracts/contracts/L1/SecurityCouncil.sol [hunk 2]
```diff
@@ -112,7 +102,9 @@ contract SecurityCouncil is MultiSigWallet, Semver {
             Colosseum.forceDeleteOutput.selector,
             _outputIndex
         );
-        uint256 transactionId = submitTransaction(COLOSSEUM, 0, message);
+        uint256 transactionId = submitTransaction(address(COLOSSEUM), 0, message);
+        // auto-confirmed by requester
+        confirmTransaction(transactionId);
```
Comment (0xHansLee) [inline packages/contracts/contracts/L1/SecurityCouncil.sol:107]: If `confirmTransaction` is included in `submitTransaction` as done in `MultiSigWallet.sol`, we don't need this.

**Review by 0xHansLee — COMMENTED** [review]
Please update `gas-snapshot` and `storage-layout`.
And why don't we merge two contract files (`SecurityCouncil.sol` and `SimpleMultiSigWallet.sol`)?

File: packages/contracts/contracts/universal/SimpleMultiSigWallet.sol [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+    /**
+     * @dev Clock (as specified in EIP-6372) is set to match the token's clock.
+     * Fallback to block numbers if the token does not implement EIP-6372.
+     */
+    function clock() public view returns (uint48) {
+        try TOKEN.clock() returns (uint48 timepoint) {
+            return timepoint;
+        } catch {
+            return SafeCastUpgradeable.toUint48(block.number);
```
[the remark this answers is not in the document]
Comment (kangsorang) [reply packages/contracts/contracts/universal/SimpleMultiSigWallet.sol:222]: The default units are block numbers, so we need to unify them.

File: packages/contracts/contracts/universal/SimpleMultiSigWallet.sol [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+     */
+    modifier confirmed(uint256 _transactionId) {
+        require(isConfirmed(_transactionId), "SimpleMultiSigWallet: not confirmed yet");
+        _;
+    }
+
+    /**
+     * @param _governor Address of the Governor contract.
+     */
+    constructor(address payable _governor) {
```
[the remark this answers is not in the document]
Comment (kangsorang) [reply packages/contracts/contracts/universal/SimpleMultiSigWallet.sol:90]: The openzeppelin GovernorUpgradable contract is payable by default. This is in case it becomes an executor.

File: packages/contracts/contracts/universal/SimpleMultiSigWallet.sol (hunk 1 above)
Comment (kangsorang) [reply packages/contracts/contracts/universal/SimpleMultiSigWallet.sol:102]: I think separating auto confirm from the behavior of submit makes the intent of the function a bit clearer, and the behavior in requestValidation is cleaner.

File: packages/contracts/contracts/universal/ISimpleMultiSigWallet.sol [hunk 5]
```diff
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.15;
+
+/**
+ * @title ISimpleMultiSigWallet
+ * @notice Interface for contracts of a simple multisig wallet.
+ */
+interface ISimpleMultiSigWallet {
```
[the remark this answers is not in the document]
Comment (kangsorang) [reply packages/contracts/contracts/universal/ISimpleMultiSigWallet.sol]: I've tried to keep it to the basic interface that our clients currently use, but there are things that need to be considered in each case, so I think it's best to put them in when they're needed.

File: packages/contracts/contracts/libraries/Types.sol [hunk 6]
```diff
@@ -174,18 +174,29 @@ library Types {
...
         bool executed;
         uint256 value;
         bytes data;
     }
 
+    /**
+     * @notice Struct representing multisig confirmation data.
+     *
+     * @custom:field confirmationCount The sum of confirmations.
+     * @custom:field confirmedBy      Map data that stores whether confirmation is performed by account.
```
Comment (0xHansLee) [inline packages/contracts/contracts/libraries/Types.sol:193]: please adjust indent here

File: packages/contracts/contracts/universal/ITokenMultiSigWallet.sol [hunk 7]
```diff
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.15;
+
+/**
+ * @title ITokenMultiSigWallet.sol
+ * @notice Interface for contracts of a simple multisig wallet.
```
Comment (0xHansLee) [inline packages/contracts/contracts/universal/ITokenMultiSigWallet.sol:6]:
```suggestion
 * @notice Interface for contracts of a token based multi-signature wallet.
```

File: packages/contracts/contracts/universal/ITokenMultiSigWallet.sol [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+ */
+interface ITokenMultiSigWallet {
```
Comment (0xHansLee) [inline packages/contracts/contracts/universal/ITokenMultiSigWallet.sol:8]: Why don't we add `quorum()` and `getVotes()` as well in interface?

File: packages/contracts/contracts/universal/ISimpleMultiSigWallet.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (0xHansLee) [reply packages/contracts/contracts/universal/ISimpleMultiSigWallet.sol]: I got your point, but I think these are needed for now. For example, even if our client is not using `revokeConfirmation` tx, this should be implemented because in some exceptional case we may need to revoke the confirmation manually.
For current contract, nobody can revoke their confirmation. It is quite weird to me. 

And it seems that we don't need `executionFailed` event because our `executeTransaction` is implemented so that the tx must succeed.

File: packages/contracts/contracts/universal/KromaSoulBoundERC721.sol [hunk 9]
```diff
@@ -106,6 +106,7 @@ abstract contract KromaSoulBoundERC721 is
         _tokenIdCounter.increment();
         _safeMint(to, tokenId);
         _setTokenURI(tokenId, uri);
+        _delegate(to, to);
```
Comment (0xHansLee) [inline packages/contracts/contracts/universal/KromaSoulBoundERC721.sol:109]: With this change, we may need to upgrade `SecurityCouncilToken` contract as well, right?
Accordingly I think we need to update semver of the contract to `v1.0.1`.

And this is out of scope of this PR I think. For the next time, let's make another PR.

File: utils/chain-ops/genesis/layer_one.go [hunk 10]
```diff
@@ -272,7 +254,8 @@ func BuildL1DeveloperGenesis(config *DeployConfig) (*core.Genesis, error) {
 	if err != nil {
 		return nil, err
 	}
-	data, err = securityCouncilTokenABI.Pack("initialize", config.SecurityCouncilTokenOwner)
+
+	data, err = securityCouncilTokenABI.Pack("initialize", opts.From)
```
Comment (0xHansLee) [inline utils/chain-ops/genesis/layer_one.go:258]: Why do you change this token owner to deployer?

File: packages/contracts/contracts/test/CommonTest.t.sol [hunk 11]
```diff
@@ -93,9 +93,89 @@ contract CommonTest is Test {
...
+                    initialProposalThreshold,
+                    votesQuorumFraction
+                )
+            )
+        );
+
+        //change proxy admin to upgradeGovernor
+        pp(address(securityCouncilToken)).changeAdmin(address(upgradeGovernor));
+        pp(address(timeLock)).changeAdmin(address(upgradeGovernor));
+        pp(address(upgradeGovernor)).changeAdmin(address(upgradeGovernor));
```
Comment (0xHansLee) [inline packages/contracts/contracts/test/CommonTest.t.sol:172]: Should the admins of these contracts be changed to timeLock? not upgradeGovernor.

File: packages/contracts/contracts/test/CommonTest.t.sol [hunk 12]
```diff
@@ -93,9 +93,89 @@ contract CommonTest is Test {
...
+    uint256 internal initialProposalThreshold = 1;
+    uint256 internal votesQuorumFraction = 70;
+    uint256 internal minDelaySeconds = 3;
+    string internal baseUri = "";
+
+    // Test data
+    address internal guardian1 = 0x0000000000000000000000000000000000001004;
+    address internal guardian2 = 0x0000000000000000000000000000000000001005;
+    address internal guardian3 = 0x0000000000000000000000000000000000001006;
+    address internal notGuardian = 0x0000000000000000000000000000000000001006;
```
Comment (0xHansLee) [inline packages/contracts/contracts/test/CommonTest.t.sol:119]: Both addresses are the same.

File: packages/contracts/contracts/test/SecurityCouncil.t.sol [hunk 13]
```diff
@@ -1,239 +1,274 @@
...
+        securityCouncilToken.safeMint(guardian2, baseUri);
+        vm.prank(owner);
+        securityCouncilToken.safeMint(guardian3, baseUri);
+
+        vm.prank(guardian1);
+        securityCouncilToken.delegate(guardian1);
+        vm.prank(guardian2);
+        securityCouncilToken.delegate(guardian2);
+        vm.prank(guardian3);
+        securityCouncilToken.delegate(guardian3);
```
Comment (0xHansLee) [inline packages/contracts/contracts/test/SecurityCouncil.t.sol:48]: We don't need to delegate token anymore? As `safeMint` also do delegation.

File: packages/contracts/contracts/test/SecurityCouncil.t.sol [hunk 14]
```diff
@@ -1,239 +1,274 @@
...
+        vm.expectRevert("TokenMultiSigWallet: onlyOwner");
+        securityCouncil.confirmTransaction(transactionId);
+        vm.stopPrank();
     }
 
-    function test_revokeConfirmation_succeeds() external {
-        // submit dummy transaction
-        vm.prank(address(owners[0]));
-        uint256 txId = securityCouncil.submitTransaction(owners[0], 0, bytes("anydata"));
+    function test_confirmTransaction_transactionExists_reverts() external {
```
Comment (0xHansLee) [inline packages/contracts/contracts/test/SecurityCouncil.t.sol:109]:
```suggestion
    function test_confirmTransaction_nonExistent_reverts() external {
```

File: packages/contracts/contracts/test/SecurityCouncil.t.sol [hunk 15]
```diff
@@ -1,239 +1,274 @@
...
-        // check confirmation revoked
-        address[] memory confirmList;
-        confirmList = securityCouncil.getConfirmations(txId);
-        assertEq(confirmList.length, 0);
+    function test_confirmTransaction_succeeds() external {
+        vm.startPrank(guardian1);
+        uint256 transactionId = securityCouncil.submitTransaction(txTarget, txValue, txData);
+        vm.expectEmit(true, true, false, true);
+        emit TransactionConfirmed(guardian1, transactionId);
+        securityCouncil.confirmTransaction(transactionId);
```
Comment (0xHansLee) [inline packages/contracts/contracts/test/SecurityCouncil.t.sol:130]: Let's check if the confirmation count is increased by the votes of the voter.

File: e2e/e2eutils/setup.go [hunk 16]
```diff
@@ -105,7 +105,11 @@ func MakeDeployParams(t require.TestingT, tp *TestParams) *DeployParams {
 		ColosseumSegmentsLengths:       "2,2,3,3",
 
 		SecurityCouncilNumConfirmationRequired: 1,
-		SecurityCouncilOwners:                  []common.Address{addresses.Challenger1, addresses.Alice, addresses.Bob, addresses.Mallory},
+		SecurityCouncilOwners: []common.Address{
+			common.HexToAddress("0x15d34AAf54267DB7D7c367839AAf71A00a2C6A65"),
+			common.HexToAddress("0x9965507D1a55bcC2695C58ba16FB37d819B0A4dc"),
+			common.HexToAddress("0x90F79bf6EB2c4f870365E785982E1f101E93b906"),
```
Comment (seolaoh) [inline e2e/e2eutils/setup.go:111]: Which address are these?

Comment (seolaoh) [reply e2e/e2eutils/setup.go:111]: Don't we need to change also in `e2e/setup.go`?

File: utils/chain-ops/genesis/testdata/test-deploy-config-devnet-l1.json [hunk 17]
```diff
@@ -38,9 +38,9 @@
   "colosseumSegmentsLengths": "3,11",
   "securityCouncilNumConfirmationRequired": 1,
```
Comment (seolaoh) [inline utils/chain-ops/genesis/testdata/test-deploy-config-devnet-l1.json:39]: We don't need this one. Please remove all variables named `SecurityCouncilNumConfirmationRequired` and `securityCouncilNumConfirmationRequired`

File: packages/contracts/contracts/universal/TokenMultiSigWallet.sol [hunk 18]
```diff
@@ -0,0 +1,214 @@
...
+        return confirmations[_transactionId].confirmationCount >= quorum();
+    }
+
+    /**
+     * @notice Returns whether the account has confirmed the transaction.
+     *
+     * @param _transactionId Transaction target address.
+     * @param _account       Address to check.
+     */
+    function hasConfirmed(uint256 _transactionId, address _account) public view returns (bool) {
```
[the remark this answers is not in the document]
Comment (seolaoh) [reply packages/contracts/contracts/universal/TokenMultiSigWallet.sol:167]: Maybe `isConfirmedBy` is the good candidate.

**Comment by seolaoh** [comment]
Additionally, we have a commit scope named `contracts`, not `contract`. Please modify it in PR title and commit msg.
Also we have agreed to start the branch name with a verb like `feat/intergrate-token-based-permissions`, please follow it next time!

File: e2e/e2eutils/setup.go (hunk 16 above)
Comment (Pangssu) [reply e2e/e2eutils/setup.go:111]: Why not use an existing address variables?

File: packages/contracts/contracts/libraries/Types.sol [hunk 19]
```diff
@@ -174,18 +174,29 @@ library Types {
     /**
      * @notice Struct representing multisig transaction data.
      *
-     * @custom:field destination The destination address to run the transaction.
+     * @custom:field target      The destination address to run the transaction.
      * @custom:field executed    Record whether a transaction was executed or not.
      * @custom:field value       The value passed in while executing the transaction.
      * @custom:field data        Calldata for transaction.
```
Comment (Pangssu) [inline packages/contracts/contracts/libraries/Types.sol:180]:
```suggestion
     * @custom:field target   The destination address to run the transaction.
     * @custom:field executed Record whether a transaction was executed or not.
     * @custom:field value    The value passed in while executing the transaction.
     * @custom:field data     Calldata for transaction.
```

indent

File: packages/contracts/contracts/test/CommonTest.t.sol [hunk 20]
```diff
@@ -93,9 +93,89 @@ contract CommonTest is Test {
             abi.encodePacked(_mint, _value, _gasLimit, _isCreation, _data)
         );
     }
+
+    //payable proxy
+    function pp(address target) internal pure returns (Proxy) {
+        return Proxy(payable(target));
+    }
```
Comment (Pangssu) [inline packages/contracts/contracts/test/CommonTest.t.sol:100]: Abbreviations don't look good.

`function toProxy()` is better.

File: packages/contracts/contracts/universal/ITokenMultiSigWallet.sol [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+     * @param transactionId Index of transaction executed.
+     */
+    event TransactionExecuted(address indexed executer, uint256 indexed transactionId);
+
+    /**
+     * @notice Allows an owner to submit and confirm a transaction.
+     *
+     * @param _target   Transaction target address.
+     * @param _value    Transaction ether value.
+     * @param _data     Transaction data payload.
```
Comment (Pangssu) [inline packages/contracts/contracts/universal/ITokenMultiSigWallet.sol:38]: indent too

File: packages/contracts/contracts/universal/TokenMultiSigWallet.sol [hunk 22]
```diff
@@ -0,0 +1,214 @@
...
+     * @notice The number of transactions submitted.
+     */
+    uint256 public transactionCount;
+
+    /**
+     * @notice Only allow the owner of governance token to call the functions.
+     *         This ensures that function is only executed by governance.
+     */
+    modifier onlyOwner(address _address) {
+        require(getVotes(_address) > 0, "TokenMultiSigWallet: onlyOwner");
```
[the remark this answers is not in the document]
Comment (Pangssu) [reply packages/contracts/contracts/universal/TokenMultiSigWallet.sol:50]: `modifier onlyTokenOwner(address _address)` is better.

File: packages/contracts/contracts/universal/TokenMultiSigWallet.sol [hunk 23]
```diff
@@ -0,0 +1,214 @@
...
+        // execute transaction if condition is met.
+        if (confirmations[_transactionId].confirmationCount >= quorum()) {
+            executeTransaction(_transactionId);
+        }
+    }
+
+    /**
+     * @inheritdoc ITokenMultiSigWallet
+     */
+    function executeTransaction(uint256 _transactionId)
```
[the remark this answers is not in the document]
Comment (Pangssu) [reply packages/contracts/contracts/universal/TokenMultiSigWallet.sol:139]: @seolaoh `executeTransaction()` don't need any restriction.

[the remark this answers is not in the document]
Comment (seolaoh) [reply packages/contracts/contracts/universal/TokenMultiSigWallet.sol:139]: Before in `MultiSigWallet.sol`, we checked if the `msg.sender` is included in owner or not. Is the policy changed?

File: utils/chain-ops/genesis/layer_one.go (hunk 10 above)
Comment (kangsorang) [reply utils/chain-ops/genesis/layer_one.go:258]: I used it because it is an EOA that can do minting in this block.

File: packages/contracts/contracts/universal/ITokenMultiSigWallet.sol [hunk 24]
```diff
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.15;
+
+/**
+ * @title ITokenMultiSigWallet.sol.sol
```
Comment (seolaoh) [inline packages/contracts/contracts/universal/ITokenMultiSigWallet.sol:5]:
```suggestion
 * @title ITokenMultiSigWallet
```

File: packages/contracts/contracts/universal/TokenMultiSigWallet.sol [hunk 25]
```diff
@@ -0,0 +1,238 @@
...
+pragma solidity 0.8.15;
+
+import "@openzeppelin/contracts-upgradeable/interfaces/IERC5805Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/math/SafeCastUpgradeable.sol";
+
+import { UpgradeGovernor } from "../governance/UpgradeGovernor.sol";
+import { SafeCall } from "../libraries/SafeCall.sol";
+import { Types } from "../libraries/Types.sol";
+import {ITokenMultiSigWallet} from "./ITokenMultiSigWallet.sol";
```
Comment (seolaoh) [inline packages/contracts/contracts/universal/TokenMultiSigWallet.sol:11]:
```suggestion
import { ITokenMultiSigWallet } from "./ITokenMultiSigWallet.sol";
```

File: packages/contracts/contracts/universal/TokenMultiSigWallet.sol [hunk 26]
```diff
@@ -0,0 +1,238 @@
...
+
+/**
+ * @custom:upgradeable
+ * @title TokenMultiSigWallet
+ * @notice This contract implements `ITokenMultiSigWallet`.
+ *         Allows multiple parties to agree on transactions before execution.
+ */
+abstract contract TokenMultiSigWallet is ITokenMultiSigWallet, ReentrancyGuardUpgradeable {
+    /**
+     * @notice The address of the governor token contract. Can be updated via upgrade.
```
Comment (seolaoh) [inline packages/contracts/contracts/universal/TokenMultiSigWallet.sol:21]:
```suggestion
     * @notice The address of the governor contract. Can be updated via upgrade.
```

File: packages/contracts/contracts/universal/TokenMultiSigWallet.sol [hunk 27]
```diff
@@ -0,0 +1,238 @@
...
+     * @inheritdoc ITokenMultiSigWallet
+     */
+    function getVotes(address account) public view returns (uint256) {
+        return IERC5805Upgradeable(address(GOVERNOR.token())).getVotes(account);
+    }
+
+    /**
+     * @notice Returns whether the account has confirmed the transaction.
+     *
+     * @param _transactionId Index of transaction to check.
```
Comment (seolaoh) [inline packages/contracts/contracts/universal/TokenMultiSigWallet.sol:201]:
```suggestion
     * @param _transactionId Transaction id to check.
```

File: packages/contracts/contracts/universal/TokenMultiSigWallet.sol [hunk 28]
```diff
@@ -0,0 +1,238 @@
...
+     * @param _account       Address to check.
+     */
+    function isConfirmedBy(uint256 _transactionId, address _account) public view returns (bool) {
+        return confirmations[_transactionId].confirmedBy[_account];
+    }
+
+    function getConfirmationCount(uint256 _transactionId) public view returns (uint256) {
```
Comment (seolaoh) [inline packages/contracts/contracts/universal/TokenMultiSigWallet.sol:208]: Please add function explanation comments

File: packages/contracts/contracts/universal/ITokenMultiSigWallet.sol [hunk 29]
```diff
@@ -0,0 +1,101 @@
...
+
+    /**
+     * @notice Allows an owner to confirm a transaction.
+     *
+     * @param _transactionId Transaction ID.
+     */
+    function confirmTransaction(uint256 _transactionId) external;
+
+    /**
+     * @notice Allows an owner to confirm a transaction.
```
Comment (seolaoh) [inline packages/contracts/contracts/universal/ITokenMultiSigWallet.sol:64]: This comment is wrong

File: packages/contracts/contracts/universal/ITokenMultiSigWallet.sol [hunk 30]
```diff
@@ -0,0 +1,101 @@
...
+/**
+ * @title ITokenMultiSigWallet.sol.sol
+ * @notice Interface for contracts of a token based multi-signature wallet.
+ */
+interface ITokenMultiSigWallet {
+    /**
+     * @notice Emitted when anyone submit a transaction.
+     *
+     * @param submitter     Owner of address that submitted a transaction.
+     * @param transactionId Index of transaction submitted.
```
Comment (seolaoh) [inline packages/contracts/contracts/universal/ITokenMultiSigWallet.sol:13]: Currently, it is not the index, unique id. Please change this comment and below comments also to transaction id.

File: packages/contracts/contracts/universal/ITokenMultiSigWallet.sol [hunk 31]
```diff
@@ -0,0 +1,101 @@
...
+     * @return Confirmation status.
+     */
+    function isConfirmed(uint256 _transactionId) external view returns (bool);
+
+    /**
+     * @notice Returns the current quorum, in terms of number of votes.
+     *
+     * @return Current quorum, in terms of number of votes: `supply * quorumNumerator / quorumDenominator`.
+     */
+    function quorum() external view returns (uint256);
```
Comment (seolaoh) [inline packages/contracts/contracts/universal/ITokenMultiSigWallet.sol:91]: Why is it `external`, not `public`?

File: packages/contracts/contracts/universal/TokenMultiSigWallet.sol [hunk 32]
```diff
@@ -0,0 +1,238 @@
...
+    /**
+     * @inheritdoc ITokenMultiSigWallet
+     */
+    function revokeConfirmation(uint256 _transactionId)
+        public
+        onlyTokenOwner(msg.sender)
+        transactionExists(_transactionId)
+    {
+        require(isConfirmedBy(_transactionId, msg.sender), "TokenMultiSigWallet: not confirmed yet");
+        require(!transactions[_transactionId].executed, "TokenMultiSigWallet: already executed");
```
Comment (seolaoh) [inline packages/contracts/contracts/universal/TokenMultiSigWallet.sol:148]: How about add `notExecuted` modifier for this function and `executeTransaction`?

File: packages/contracts/contracts/universal/ISimpleMultiSigWallet.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (seolaoh) [reply packages/contracts/contracts/universal/ISimpleMultiSigWallet.sol]: `getTransactionCount`, `getConfirmations` is still missing.

File: packages/contracts/contracts/universal/TokenMultiSigWallet.sol [hunk 33]
```diff
@@ -0,0 +1,238 @@
...
+        bytes memory _data
+    ) public view validAddress(_target) returns (uint256) {
+        return uint256(keccak256(abi.encode(_target, _value, _data, clock())));
+    }
+
+    /**
+     * @dev Clock (as specified in EIP-6372) is set to match the token's clock.
+     *      Fallback to block numbers if the token does not implement EIP-6372.
+     */
+    function clock() public view returns (uint48) {
```
Comment (seolaoh) [inline packages/contracts/contracts/universal/TokenMultiSigWallet.sol:231]: I wonder why these functions are not included in the interface.

Comment (kangsorang) [reply packages/contracts/contracts/universal/TokenMultiSigWallet.sol:231]: isConfirmedBy : added
getConfirmationCount : added
generateTransactionId : This is not a function used in the general case.
clock : This is not a function used in the general case.

**Review by seolaoh — APPROVED** [review]

**Review by 0xHansLee — COMMENTED** [review]
Please update the semver of `KromaSoulBoundERC721.sol` to `v1.0.1`.

File: packages/contracts/contracts/universal/TokenMultiSigWallet.sol [hunk 34]
```diff
@@ -0,0 +1,253 @@
...
+        txn.executed = true;
+        bool success = SafeCall.call(txn.target, gasleft(), txn.value, txn.data);
+        require(success, "TokenMultiSigWallet: call transaction failed");
+        emit TransactionExecuted(msg.sender, _transactionId);
+    }
+
+    /**
+     * @inheritdoc ITokenMultiSigWallet
+     */
+    function isConfirmed(uint256 _transactionId) public view returns (bool) {
```
[the remark this answers is not in the document]
Comment (kangsorang) [reply packages/contracts/contracts/universal/TokenMultiSigWallet.sol:190]: I'd love to, but since guardian.go is already implemented to use isConfirmed to mean checking for excusability, I can't rename it for compatibility.

**Review by Pangssu — APPROVED** [review]

**Review by 0xHansLee — APPROVED** [review]
LGTM 🚀
